Source/WebCore/dom/Document.cpp

```diff
@@ -9736,6 +9741,37 @@ std::optional<uint64_t> Document::noiseInjectionHashSalt() const
     return page()->noiseInjectionHashSaltForDomain(RegistrableDomain { m_url });
 }
 
+ContentVisibilityDocumentState& Document::contentVisibilityDocumentState()
+{
+    if (!m_contentVisibilityDocumentState)
+        m_contentVisibilityDocumentState = makeUnique<ContentVisibilityDocumentState>();
+    return *m_contentVisibilityDocumentState;
+}
+
+bool Document::isObservingContentVisibilityTargets() const
+{
+    return m_contentVisibilityDocumentState && m_contentVisibilityDocumentState->hasObservationTargets();
+}
+
+void Document::updateRelevancyOfContentVisibilityElements()
+{
+    if (!isObservingContentVisibilityTargets())
+        return;
+    if (m_contentVisibilityDocumentState->hasSameFrameObservationTargets()) {
+        updateIntersectionObservations();
+        m_contentVisibilityDocumentState->clearSameFrameObservationTargets();
+    }
```

Comment:
Not super convinced this is the best way to handle this, I'd be curious to know what other browsers do

Reply:
Yes, this part is tricky since it is not specified well. The situation is described well here:
https://github.com/w3c/csswg-drafts/issues/8542

Basically IIUC chromium makes this work by using sync intersection observer internally, which is not specced AFAIK and not something gecko or WebKit provide. Therefore chromium is the only supporting this, it is explained in https://github.com/w3c/csswg-drafts/issues/8542#issuecomment-1459175417

I propose to leave out this support in this PR and open a bug for it, fixing it once the issue is a bit more discussed/accepted. Note that the only test it has an effect on for the test result is auto-012.html, it conveniently would "fix" the WK1 issues but the real fix for that probably still is https://bugs.webkit.org/show_bug.cgi?id=259075.

Reply:
> I propose to leave out this support in this PR and open a bug for it, fixing it once the issue is a bit more discussed/accepted. Note that the only test it has an effect on for the test result is auto-012.html,

This test seems to pass in Gecko, so it would be nice seeing why that is the case.

But I'm OK implementing the Chromium proposal separately though.